presto-spi/src/main/java/com/facebook/presto/spi/block/DictionaryBlock.java

```diff
@@ -335,22 +338,20 @@ public DictionaryBlock compact()
             return blocks;
         }
 
-        Slice newIdsSlice = wrappedIntArray(getNewIds(positionCount, ids, remapIndex));
-        List<Block> outputDictionaryBlocks = new ArrayList<>(blocks.size());
-        DictionaryId dictionaryId = randomDictionaryId();
-
-        for (Block block : blocks) {
-            dictionaryBlock = ((DictionaryBlock) block);
-            try {
-                Block compactDictionary = dictionaryBlock.getDictionary().copyPositions(dictionaryPositionsToCopy);
-                outputDictionaryBlocks.add(new DictionaryBlock(positionCount, compactDictionary, newIdsSlice, true, dictionaryId));
-            }
-            catch (UnsupportedOperationException e) {
-                // ignore if copy positions is not supported for the dictionary
-                outputDictionaryBlocks.add(new DictionaryBlock(positionCount, dictionaryBlock.getDictionary(), dictionaryBlock.getIds()));
-            }
+        Block compactDictionary;
+        try {
+            compactDictionary = commonDictionaryBlock.getDictionary().copyPositions(dictionaryPositionsToCopy);
         }
-        return outputDictionaryBlocks;
+        catch (UnsupportedOperationException e) {
+            // ignore if copy positions is not supported for the dictionary
+            return blocks;
+        }
+
+        DictionaryId dictionaryId = randomDictionaryId();
+        return blocks.stream().map(DictionaryBlock.class::cast).map(dictionaryBlock -> {
+            Slice newIdsSlice = wrappedIntArray(getNewIds(dictionaryBlock.positionCount, dictionaryBlock.ids, remapIndex));
+            return new DictionaryBlock(dictionaryBlock.positionCount, compactDictionary, newIdsSlice, true, dictionaryId);
```

Comment:
I have left it because it already existed in the existing implementation. If a dictionary is shared between blocks, then almost all blocks will get `false`. If someone calls `compact ` on such a block in a later step, `n` compacted copies would be created for the shared dictionary. Therefore, in a later step, I wanted to introduce a state `shared` which counts as `compact`.